driver-core/src/main/com/mongodb/internal/connection/BaseCluster.java

```diff
@@ -477,4 +495,78 @@ public void run() {
             }
         }
     }
+
+    private void logServerSelectionStarted(
+            final OperationContext operationContext,
+            final ServerSelector serverSelector,
+            final ClusterDescription clusterDescription) {
+        if (STRUCTURED_LOGGER.isRequired(DEBUG, clusterId)) {
+            STRUCTURED_LOGGER.log(new LogMessage(
+                    SERVER_SELECTION, DEBUG, "Server selection started", clusterId,
+                    asList(
+                            new Entry(OPERATION, null),
+                            new Entry(OPERATION_ID, operationContext.getId()),
+                            new Entry(SELECTOR, serverSelector.toString()),
+                            new Entry(TOPOLOGY_DESCRIPTION, clusterDescription.getShortDescription())),
+                    "Server selection started for operation[ {}] with ID {}. Selector: {}, topology description: {}"));
+        }
+    }
+
+    private void logServerSelectionWaiting(
+            final OperationContext operationContext,
+            @Nullable
+            final Long remainingTimeNanos,
+            final ServerSelector serverSelector,
+            final ClusterDescription clusterDescription) {
+        if (STRUCTURED_LOGGER.isRequired(INFO, clusterId)) {
+            STRUCTURED_LOGGER.log(new LogMessage(
+                    SERVER_SELECTION, INFO, "Waiting for suitable server to become available", clusterId,
+                    asList(
+                            new Entry(OPERATION, null),
+                            new Entry(OPERATION_ID, operationContext.getId()),
+                            new Entry(REMAINING_TIME_MS, remainingTimeNanos == null ? null : NANOSECONDS.toMillis(remainingTimeNanos)),
+                            new Entry(SELECTOR, serverSelector.toString()),
+                            new Entry(TOPOLOGY_DESCRIPTION, clusterDescription.getShortDescription())),
+                    "Waiting for server to become available for operation[ {}] with ID {}.[ Remaining time: {} ms.]"
+                            + " Selector: {}, topology description: {}."));
+        }
+    }
+
+    private void logServerSelectionFailed(
+            final OperationContext operationContext,
+            final MongoException failure,
+            final ServerSelector serverSelector,
+            final ClusterDescription clusterDescription) {
+        if (STRUCTURED_LOGGER.isRequired(DEBUG, clusterId)) {
+            STRUCTURED_LOGGER.log(new LogMessage(
+                    SERVER_SELECTION, DEBUG, "Server selection failed", clusterId,
+                    asList(
+                            new Entry(OPERATION, null),
+                            new Entry(OPERATION_ID, operationContext.getId()),
+                            new Entry(FAILURE, failure.toString()),
```

Comment:
I know it's in the specification, but it's not clear what useful information can be included in the failure string.   As far as I can see it's always the same message.  @BorisDog can you clarify the intention here?

Reply:
[Spec](https://github.com/mongodb/specifications/blob/master/source/server-selection/server-selection.rst#server-selection-errors) defines different errors in different cases, mainly:
1. Timeout
2. Incompatibility (few flavors)